### ISSUE TITLE
generator: more idiomatic language APIs for errors and optional values

### ISSUE THREAD

**Issue by LeSeulArtichaut** [prologin/stechec2#159]
In this year's Prologin finals, I had discussions with fellow rustaceans on how the generated Rust API felt clumsy and unidiomatic at times. We concluded that it would be great if the API could better match the language's idioms and common patterns more.

## The problem

Here are two common patterns that would benefit greatly from a better API:

### Optional values

Sometimes, fetching some information through an API call can fail, e.g. if the provided input is invalid. In this case, games usually return a sentinel value that represents "no meaningful value", like `-1`. This might be idiomatic in C, but since optional values are literally everywhere in programming, modern (i.e. non-C) languages have better ways of representing optional values:
- Rust, OCaml and Haskell have `Option`, `option` and `Maybe` respectively
- Python, Java, C# and PHP all have some flavor of `null`

### Errors

Similarly, a champion might attempt to perform an invalid action, in which case the came usually responds with an error enumeration giving more details on why the attempted action was invalid. Once again, every language has its own error handling mechanisms (also see #89):
- Rust has `Result`, OCaml has `result`, Haskell has `Exception`*
- Python, Java, C# and PHP have exceptions*
- C++ programmers can't agree on how to handle errors, leave them alone

\* There is a caveat here: unhandled exceptions are propagated and can crash the program, and developers might be forced to write code to handle them (e.g. Java). Competitive programming is not the place for proper error handling and beautiful error messages. There is a trade-off here: having errors crash the program is good for debugging (no error goes unnoticed), but bad for tournaments where you want to continue the match no matter what, and bad for writing quick and dirty code.

## Proposed solution

The API is built on top of a C FFI layer, which can hardly be changed as it needs to be common to all languages. However, nothing prevents us from building language-specific abstractions on top of the FFI.
I think the two cases presented in this issue are fundamental and deserve special treatment. If we consider this to be true, then we can integrate them in the YAML config files used for specifying a game's API.

### Optional values

Functions can mark a function's return value as optional by specifying the sentinel value in a `fct_ret_none_val` field. This tells the generator to generate code to transform the sentinel into the langage's idiomatic way of representing missing values, like `None` in Rust or `null` in Java.

### Errors

Errors are specified in a new top-level list `error` (same level as `enum` or `function`), which has fields `err_name`, `err_summary`, `err_kind`. `err_kind` follows the same format as `enum_field`.

At the C FFI level, the error is represented as an `enum`, where the first variant (value `0`) is an `OK` variant representing success.

Language APIs now translate this C enum into the language's idiomatic exception mechanism.

---
I can work on implementing this myself, but I would like some feedback on whether this is useful (though I have no doubt it is), and whether this is the right way to go. cc @remi-dupre

**Comment by seirl** [comment]
Yes please!

This has been on our wishlist for ages. There is https://github.com/prologin/stechec2/issues/143 to fix as a dependency of this task.

**Comment by Angramme** [comment]
+1 
pls pls pls

**Comment by Angramme** [comment]
ohhhhhh,

this sounds really sexy

**Comment by seirl** [comment]
Two other things come to mind:

- We need to decide the exact semantic before implementing anything. What happens when you try to write a function that takes an error as an argument? Can you have arrays of errors? Should errors really be a type or should they be in a `throws: ` field of a function description instead? Do we require functions that can throw errors to return void for compatibility with the C interface?

- There are a couple of things that are currently broken with the generators, like tuples in haskell. We need to make the generators more robust in their current capabilities and add tests to ensure we're not breaking too much stuff when adding more complex semantics.

**Comment by 71** [comment]
While I agree with most things said by @LeSeulArtichaut, I have a few comments:
- OCaml and C++ have result types, but use exceptions more extensively, while they both have [`Option`](https://ocaml.org/api/Option.html) / [`std::optional`](https://en.cppreference.com/w/cpp/utility/optional) types. Java and C# also fit here, but with `null`.
- Haskell also has `Either`, which AFAIK is used quite a lot. I'm not very familiar with it, but wrapping the whole game state in a monad (with Exception and State) would be another alternative.
  - On the topic of state, Rust (with a `Default` struct) and possibly C++ (with a default constructible class) would benefit from having non-global state.
- I believe there should be different error types. For the last finale, only two values of the `erreur` enum (`ok` and `id_panda_invalide`, which was actually never used) could be shared among actions. All the others were specific to one of the actions.
  - For more advanced errors as mentioned by @seirl, if we get there, we could have our own "result" type for C by generating structs, enums and functions individually for each result type.

Regarding the caveat of having exceptions throw: there are already many ways to make a program crash (for instance, by indexing into an array that is too small). The API is indeed much more likely to fail due to an error, and of the code I've seen, in many cases the errors are printed to stdout or simply ignored, which is different from regular errors like null pointer deferences. I thus generally agree that we should strive to have non-fatal errors as much as possible, if we want to avoid having `try catch` statements all over the place. The languages I mentioned above that use exceptions more extensively than result types **do** support error types (in the sense that we could provide our own error type, or use a library). In my opinion, we could provide both an idiomatic function with exceptions, and another one with a result type, though this may needlessly complexify the API.

**Comment by LeSeulArtichaut** [comment]
> What happens when you try to write a function that takes an error as an argument?

I really don't see any case where passing errors as function arguments would be useful, and implementing it would needlessly make the generator more complex. I think it should be invalid at the configuration level.

> Can you have arrays of errors?

I think this is rarely useful. If an error occurs, the most common action to take is to just return early, so you can't have multiple errors happening. Ultimately this depends on how we declare errors. Speaking of...

> Should errors really be a type or should they be in a `throws:` field of a function description instead?

*/me grabs a cup of Java coffee* :coffee:

This is interesting, because it forces every function to declare exactly how it can fail, which, the more I think about it, is absolutely awesome for an API consumer and leads to self-documenting code. OTOH this means that defining the API takes more work, the FFI itself needs to become more complex because there needs to be an underlying error type for every fallible function, and error handling in C(++) can get more tedious because of this plurality of error types. IMO this is still worth it =)

> Do we require functions that can throw errors to return void for compatibility with the C interface?

Yep, the C FFI is limiting us here, but if we think being able to return non-`void` values in fallible functions is useful, we can probably work around those limitations. I don't know much about C, but I guess one way to go would be to return some kind of
```c
struct Error {
    ErrorKind error,  // the error kind, or `OK` on success
    ReturnType value, // the value on success
}
```
I just hope that we can return uninitialized memory as `value` when an error occurs, because I guess returning a properly initialized type is going to be a nightmare.

> There are a couple of things that are currently broken with the generators, like tuples in haskell. We need to make the generators more robust in their current capabilities and add tests to ensure we're not breaking too much stuff when adding more complex semantics.

Are there any issues detailing this problem? I guess #151?

**Comment by seirl** [comment]
I think we agree on all the major points:

- errors should be separate from types, and you shouldn't be able to have complex types or func parameters including errors
- using `throws` is better than using the return type for this
- we should (at least at first) ban functions that can throw but that do not return void.

> > There are a couple of things that are currently broken with the generators, like tuples in haskell. We need to make the generators more robust in their current capabilities and add tests to ensure we're not breaking too much stuff when adding more complex semantics.
> 
> Are there any issues detailing this problem? I guess #151?

Yes, I'd be happy to review a PR fixing this issue. Basically just try to compile the player output of all the old games from prologin2012 to now, some of them will fail.

**Comment by seirl** [comment]
Thanks for the fix in #151! I also added #161 and #162 to track potential improvements that would help consolidate the current codebase.

**Comment by rambip** [comment]
Great idea !
To give ideas about what could be a great idiomatic api for rust, here is a documented interface I created for the 2021 final edition of the contest:

https://pastebin.com/HhpTYB7J

I hope this can be usefull !
And a lot of the ideas behind this interface can be used in many languages I think.

**Comment by LeSeulArtichaut** [comment]
Trying to start somewhere: if we need a way to define a sentinel value that represents `none`, there should be a way to define constants of arbitrary types. Opened issue #169 for this.

**Comment by seirl** [comment]
Final brainstorming session with @LeSeulArtichaut :

1. For functions that only return an error, we shouldn't need to change anything except a `fct_ret_error: true` which indicates that the returned enum is supposed to be interpreted as an error. The languages like rust that support `Result`, this would add some sugar to use it. Nothing would change for the other languages.
2. Returning a value OR an error would complexify the API for low level languages, not a priority to implement. Could be discussed after the simple case is implemented.
3. For functions that can return nothing (option types/Maybe/Option/...), we can add a field `fct_none_sentinel_value:` that points to an already defined constant. For languages that support it, it would check that the value returned by the low level API is equal to this sentinel. If that's the case, it would return the None value instead of this sentinel. Depends on #169 .

**Comment by seirl** [comment]
#143 would also be solved with a `fct_ret_error: true` field.

**Comment by juli0z** [comment]
Migrated to https://gitlab.com/prologin/tech/tools/stechec2/-/issues/159
